### PR TITLE
OpenGL 2.1 support

### PR DESCRIPTION
**Pull request by fluffyfreak**
# OpenGL 2.1
An alternative Opengl 2.1 renderer in place of the standard Opengl 3.x (_'x' because of extensions etc_) for the following reasons:
- compatibility with older hardware
- improved separation between renderer & engine
- ...I saw that 'jumpdrive' did it and I wanted it...

# Missing bits
I couldn't get all of the Opengl 3.x stuff to work exactly the same in Opengl 2.1 so there are visual differences and some performance differences like:
- No GPU gas giant texture generation
- No Instanced rendering
- Starfield looks strange as points don't appear to resize correctly
- No Ansiotropic rendering
- Probably other bugs / visual issues

# Nasty bits
No doubt some of the rendering changes will have errors and problems.

# Testing:
Build the branch, run the game once, it will add a new line to the `config.ini` which will by default be set too `RendererName=Opengl 3.x`.
Change that line to read `RendererName=Opengl 2.1` and run Pioneer again.

# Summary
This should allow the game to run lower end GPU's in laptops that cannot be upgraded.
Although due to the lack of instanced rendering in the 2.1 branch it will perform much worse around cities.
Also there are missing rendering options etc.

### CODE REVIEW

**Comment by impaktor** [comment]
```
RendererGL.cpp: In constructor ‘Graphics::RendererOGL::RendererOGL(Graphics::WindowSDL*, const Graphics::Settings&)’:
RendererGL.cpp:105:56: warning: use of old-style cast [-Wold-style-cast]
  const char *ver = (const char *)glGetString(GL_VERSION);
                                                        ^
RendererGL.cpp:79:9: warning: unused variable ‘err’ [-Wunused-variable]
  GLenum err = glGetError();
         ^
In file included from RendererGL.cpp:12:0:
TextureGL.h: In member function ‘virtual Graphics::Texture* Graphics::RendererOGL::CreateTexture(const Graphics::TextureDescriptor&)’:
TextureGL.h:29:2: error: ‘Graphics::OGL::TextureGL::TextureGL(const Graphics::TextureDescriptor&, bool, bool)’ is private
  TextureGL(const TextureDescriptor &descriptor, const bool useCompressed, const bool useAnisoFiltering);
  ^
RendererGL.cpp:944:90: error: within this context
  return new OGL::TextureGL(descriptor, m_useCompressedTextures, m_useAnisotropicFiltering);
                                                                                          ^
In file included from RendererGL.cpp:12:0:
TextureGL.h: In member function ‘virtual Graphics::RenderTarget* Graphics::RendererOGL::CreateRenderTarget(const Graphics::RenderTargetDesc&)’:
TextureGL.h:29:2: error: ‘Graphics::OGL::TextureGL::TextureGL(const Graphics::TextureDescriptor&, bool, bool)’ is private
  TextureGL(const TextureDescriptor &descriptor, const bool useCompressed, const bool useAnisoFiltering);
  ^
RendererGL.cpp:979:68: error: within this context
   OGL::TextureGL *colorTex = new OGL::TextureGL(cdesc, false, false);
                                                                    ^
In file included from RendererGL.cpp:12:0:
TextureGL.h:29:2: error: ‘Graphics::OGL::TextureGL::TextureGL(const Graphics::TextureDescriptor&, bool, bool)’ is private
  TextureGL(const TextureDescriptor &descriptor, const bool useCompressed, const bool useAnisoFiltering);
  ^
RendererGL.cpp:993:69: error: within this context
    OGL::TextureGL *depthTex = new OGL::TextureGL(ddesc, false, false);
```

**Comment by fluffyfreak** [comment]
@impaktor and that's despite it being a friend class?

**Comment by impaktor** [comment]
Sorry:
```
  CXX      RendererGL.o
RendererGL.cpp: In constructor ‘Graphics::RendererOGL::RendererOGL(Graphics::WindowSDL*, const Graphics::Settings&)’:
RendererGL.cpp:105:56: warning: use of old-style cast [-Wold-style-cast]
  const char *ver = (const char *)glGetString(GL_VERSION);
                                                        ^
RendererGL.cpp:79:9: warning: unused variable ‘err’ [-Wunused-variable]
  GLenum err = glGetError();
         ^
In file included from RendererGL.cpp:12:0:
TextureGL.h: In member function ‘virtual Graphics::Texture* Graphics::RendererOGL::CreateTexture(const Graphics::TextureDescriptor&)’:
TextureGL.h:31:4: error: ‘Graphics::OGL::TextureGL::TextureGL(const Graphics::TextureDescriptor&, bool, bool)’ is private
    TextureGL(const TextureDescriptor &descriptor, const bool useCompressed, const bool useAnisoFiltering);
    ^
RendererGL.cpp:944:90: error: within this context
  return new OGL::TextureGL(descriptor, m_useCompressedTextures, m_useAnisotropicFiltering);
                                                                                          ^
In file included from RendererGL.cpp:12:0:
TextureGL.h: In member function ‘virtual Graphics::RenderTarget* Graphics::RendererOGL::CreateRenderTarget(const Graphics::RenderTargetDesc&)’:
TextureGL.h:31:4: error: ‘Graphics::OGL::TextureGL::TextureGL(const Graphics::TextureDescriptor&, bool, bool)’ is private
    TextureGL(const TextureDescriptor &descriptor, const bool useCompressed, const bool useAnisoFiltering);
    ^
RendererGL.cpp:979:68: error: within this context
   OGL::TextureGL *colorTex = new OGL::TextureGL(cdesc, false, false);
                                                                    ^
In file included from RendererGL.cpp:12:0:
TextureGL.h:31:4: error: ‘Graphics::OGL::TextureGL::TextureGL(const Graphics::TextureDescriptor&, bool, bool)’ is private
    TextureGL(const TextureDescriptor &descriptor, const bool useCompressed, const bool useAnisoFiltering);
    ^
RendererGL.cpp:993:69: error: within this context
    OGL::TextureGL *depthTex = new OGL::TextureGL(ddesc, false, false);
```

**Comment by impaktor** [comment]
Yep, compiles now!

**Comment by impaktor** [comment]
...too bad it segfaults
```
orbital_station_2-5k has:
 5 entrances,
 64 pads,
 5 exits
orbital_station_2-5k10k has:
 5 entrances,
 64 pads,
 5 exits
NavLights::Init
Sfx::Init
Sound::Init
[New Thread 0x7fffe966e700 (LWP 15451)]


Loading took: 9208.370419 milliseconds
CreateCollisionMesh for : (bluenose)

Program received signal SIGSEGV, Segmentation fault.
0x000000004009dc2a in ?? ()
(gdb) bt
#0  0x000000004009dc2a in ?? ()
#1  0x00007ffff44afd0a in ?? () from /usr/lib64/libnvidia-glcore.so.340.96
#2  0x00007ffff44b83a7 in ?? () from /usr/lib64/libnvidia-glcore.so.340.96
#3  0x00007ffff4121e68 in ?? () from /usr/lib64/libnvidia-glcore.so.340.96
#4  0x0000000000792bac in ImGui_ImplSdl_RenderDrawLists (draw_data=0x1541d70 <GImDefaultContext+6288>)
    at examples/sdl_opengl2_example/imgui_impl_sdl.cpp:83
#5  0x000000000077d90c in ImGui::Render () at imgui.cpp:2741
#6  0x00000000004f1b7c in RenderImGui () at PiGui.h:45
#7  Pi::DrawPiGui (delta=0, handler=...) at Pi.cpp:1688
#8  0x00000000004f71f8 in Pi::Start () at Pi.cpp:1251
#9  0x0000000000416239 in main (argc=<optimized out>, argv=<optimized out>) at main.cpp:130
(gdb) 
```

**Comment by fluffyfreak** [comment]
Well that's odd, it work fine in Windows, but yes I get it on my machine too :(

**Comment by fluffyfreak** [comment]
@impaktor can you give that latest change a go? There was some stupid going on, it now runs for me after that first warning. It's still spewing warnings and glErrors internally but ... closer now :)

**Comment by impaktor** [comment]
works, as far as I could tell. No crash after 2-4 minutes of play.

**Comment by fluffyfreak** [comment]
@impaktor Ok I have just done some crazy shit to the above. It should now fallback to OpenGL 2.1 by default, I have also reproduced the warning on Windows and I think that I have fixed it in https://github.com/pioneerspacesim/pioneer/pull/3934/commits/7caa990b8329a4ebb6796df5033ffd454d51c3e1 so if you wouldn't mind testing it for me that would be great!

**Comment by fluffyfreak** [comment]
This works on a heigher end OpenGL 3.x capable GPU running with our 2.1 renderer, however I don't have any real GL2.1 hardware to test it on.

**Comment by impaktor** [comment]
> ...however I don't have any real GL2.1 hardware to test it on.

...neither have I, I think my drivers were GL1.4 or something, the last time we/I/you tried going down the 2.1 branch-road. (Thinkpad X201).

Anything specific you want me to do?

**Comment by laarmen** [comment]
... Buy an newer old laptop?...

Allright, I know the way out ;-)

**Comment by mike-f1** [comment]
...ping https://github.com/paulbilkis?

**Comment by impaktor** [comment]
Works!

**Comment by mike-f1** [comment]
@fluffyfreak : I fear there's something wrong with history...

**Comment by fluffyfreak** [comment]
@mike-f1 how so?

**Comment by mike-f1** [comment]
If I'm not taking *sun for darkness*, I think you forgot to rebase, so some commit you did locally has been merged, here some ref and my git log --oneline output:

e8f653e; df0e7e7; 0751588; e6806c8... Nothing so bad, I only would show now I know how git works ;)
```

ab24333 Update Changelog.txt
20795e5 Merge pull request #3934 from fluffyfreak/glew-gl2
cc9db5a Update Changelog.txt
88952cd Merge pull request #3938 from fluffyfreak/mismatches
e8f653e Resolve the mismatched attribute numbering.
dfff95b Update Changelog.txt
aa23d59 Merge pull request #3937 from salborrelli/osx_pkg1703
9ccee08 Update Changelog.txt
60c4b70 Merge pull request #3936 from mike-f1/mv-ortho
a20ffd2 Upgrade of Xcode project for Mac OSX build
7caa990 Overwrite the ImplSdl version with the ImplSdlGL3 version and modify for use with OpenGL 2.1
df0e7e7 Fallback to OpenGL 2.1 automatically if 3.x context fails.
9dd594d Close #2665: Orthographic view in modelviewer
f2ebd5f Use the correct CheckRenderErrors function to de-duplicate code. Use the correct OpenGL headers (ours).
5b6d2e2 Link against the libgraphicsgl2 lib we've now added.
0243251 Update the imgui Makefile.am to include the OpenGL2 code

```

**Comment by laarmen** [comment]
Just reading the commit log, I don't see any problem. The "Update
Changelog.txt" commits are all different ones despite the very very very
similar shortlog. :). Also, the commits you linked to don't show up in the
quoted logs except for the first.

**Comment by mike-f1** [comment]
@laarmen : well, you're right: I snip too much from the output of git log and I explain that very badly...
But I can find the commit I cited in master: just copy the version number while browsing master and you can find all of them... Let me try again ;)

screenshot_2017-03-05_21-21-58

...to me these seems "partial commits"... But I just want to show them, no problem if you want keep them.

**Comment by laarmen** [comment]
Actually, to me this history looks rather good. The commits are small,
relatively self-contained which would make review easier. The only one that
could be argued against IMO is the one fixing the Linux Makefile, but other
than that, this history is clean.

Personally, my work flow is to do a bunch of work, with perhaps temporary
commits, and when it's all done I remove all the temporary commits, and
I'll create smaller patches with proper logs. So, I'll usually have more
commits in my PR than I did on my work-in-progress branch.

**Comment by mike-f1** [comment]
Yes: me too, but here I felt temporary commits weren't removed... Probably I'm wrong, so for it's ok also as left these as they are.

**Comment by fluffyfreak** [comment]
We've not cared that much about the commit history in the past. Or at least, I haven't :D

I know @impaktor and @laarmen would like me to keep the history cleaner but I'm really only interested in the code itself: Does it work? How can I tell it works? What does it do?

I'd like to see separate commits for each feature/fix/item so that if a bug fix has become mixed into a feature they can be cherry picked out and taken separately.

Also rewriting a Git history might work these days but I have never had a rebase work in Windows :(
